dashboard/src/shared/ResourceRef.ts

```diff
@@ -41,6 +41,22 @@ class ResourceRef {
     return Kube.getResourceURL(this.apiVersion, this.resourcePlural(), this.namespace, this.name);
   }
 
+  public watchResourceURL() {
```

Comment:
this methods are no longer related to a "resource reference", since they are wrapping `Kube` methods, should we just use those? Maybe is `Kube` who should call methods from here when it needs to resolve URLs or plurals

Reply:
Yeah this is why I was thinking about moving around the methods. However I ultimately decided against merging these as we will need a way to make arbitrary requests to the Kubernetes API and Kube.ts allows us to do that right now. These are kind of just convenience wrappers of those calls to make it easier if you are using ResourceRef. That's the thinking, but if you have any other ideas about how to structure this, I'm certainly open to it!

I'm not sure what you mean by they are not related to the ResourceRef, they just handle taking the parameters in the ResourceRef object and passing them appropriately to Kube.ts to avoid having to type out each parameter every time.

Reply:
My concern about the naming is small but what I mean is that if I read `reference.get()` I assume that it will return the `reference` object, not the resource, that's why I think this could be called only `Resource` since `Resource.get` or `Resource.watch` really returns the resource.

Apart from that, I think that for make it more clear that `Kube` is the low level library for this (and AFAIK now it should not be used directly) we can add it inside this component: `shared/Resource/Kube`.

It's also minor but I would move the `resourcePlural` method to `Kube`, it shouldn't be use externally, isn't it?

Reply:
> My concern about the naming is small but what I mean is that if I read reference.get() I assume that it will return the reference object, not the resource, that's why I think this could be called only Resource since Resource.get or Resource.watch really returns the resource.

I think I understand what you're saying here, but the problem is we don't have a `Resource` type here, we have a reference that is used to get the resource. We could have a `Resource` type that holds partial data about the GVK, namespace and name and implement `get()` to fetch the resource from the server, but it will be a huge change to use everywhere and I'm not sure that is valuable over what we currently have. IMO, `ResourceRef.getResource()` is fairly clear that it is to get the referenced resource.

> Apart from that, I think that for make it more clear that Kube is the low level library for this (and AFAIK now it should not be used directly) we can add it inside this component: shared/Resource/Kube.

I disagree that Kube should be a low-level library that is not directly used, as previously stated there will be times when we need to use Kube directly (e.g. to make a List call), when we don't have a ResourceRef. I think we will also want to eventually add support for POST/PUT/DELETE so we can use Kube.ts as the library for the calls made from AppRepository.ts, for example, to help us reduce the duplication.

> It's also minor but I would move the resourcePlural method to Kube since it shouldn't be use externally, isn't it?

Good point, I do think this might make sense to move to Kube, especially since we might need to use it for List and other calls. Assuming ResourceRef is going to call `resourcePlural` though, it will need to be public.

Reply:
> I think I understand what you're saying here, but the problem is we don't have a `Resource` type here, we have a reference that is used to get the resource. We could have a `Resource` type that holds partial data about the GVK, namespace and name and implement `get()` to fetch the resource from the server, but it will be a huge change to use everywhere and I'm not sure that is valuable over what we currently have. IMO, `ResourceRef.getResource()` is fairly clear that it is to get the referenced resource.

What I meant was simply renaming `ResourceRef` for `Resource` so `s/ResourceRef.getResource/Resource.get/`. It's not that `ResourceRef.getResource` is not clear what's doing it's just that I think the naming is a bit more natural the other way. In any case if you have a strong preference you can keep this naming.

> I disagree that Kube should be a low-level library that is not directly used, as previously stated there will be times when we need to use Kube directly (e.g. to make a List call), when we don't have a ResourceRef. I think we will also want to eventually add support for POST/PUT/DELETE so we can use Kube.ts as the library for the calls made from AppRepository.ts, for example, to help us reduce the duplication.

I thought that `Kube` was no longer used externally, if it's still used we should keep it accessible. My concern is how we make clear that when creating a new component (for example something like `JobsTable`) it should use this `ResourceRef` instead of directly calling the methods of `Kube`.

> Good point, I do think this might make sense to move to Kube, especially since we might need to use it for List and other calls. Assuming ResourceRef is going to call `resourcePlural` though, it will need to be public.

:+1:

Reply:
> What I meant was simply renaming `ResourceRef` for `Resource` so `s/ResourceRef.getResource/Resource.get/`. It's not that `ResourceRef.getResource` is not clear what's doing it's just that I think the naming is a bit more natural the other way. In any case if you have a strong preference you can keep this naming.

The problem with this is that `Resource` should have the schema of `IResource`. We could do that, but it'd be changing a lot of things if we wanted to use `Resource` everywhere instead of `IResource`. I like having a smaller structure of `ResourceRef` being used to get a full `IResource`, but I do also agree that having a single `Resource` object that might be able to get itself could be a cleaner approach overall. That said, I think it's out of the scope of this PR.

> I thought that `Kube` was no longer used externally, if it's still used we should keep it accessible. My concern is how we make clear that when creating a new component (for example something like `JobsTable`) it should use this `ResourceRef` instead of directly calling the methods of `Kube`.

It isn't currently used outside of `ResourceRef`, but my point is that it could easily be due to `ResourceRef`'s limitations (i.e. only pointing to a single resource). You have a good point about making it more clear that we should use ResourceRef instead where we can. I'll add a comment to Kube.ts to reflect this.

Reply:
> The problem with this is that `Resource` should have the schema of `IResource`. We could do that, but it'd be changing a lot of things if we wanted to use `Resource` everywhere instead of `IResource`.

That's not exactly what I was thinking. I'd be okay having the interface `IResource` and the class `Resource`, needing the constructor of `Resource` an `IResource` (just like the `ResourceRef` now).

> It isn't currently used outside of `ResourceRef`, but my point is that it could easily be due to `ResourceRef`'s limitations (i.e. only pointing to a single resource)

I thought that in that case we should extend `ResourceRef` instead of using directly `Kube` to support list/other resources, isn't it?

Reply:
As discussed in the standup, I don't have a strong preference either way on this, I would say that a "Resource" class feels like it should have the properties of a resource, and not just be used to get an "IResource" - but perhaps this comes from my Rails models days ;).

> I thought that in that case we should extend `ResourceRef` instead of using directly `Kube` to support list/other resources, isn't it?

Potentially, but maybe it's simpler to keep ResourceRef to just reference a single resource? I'm not sure, but I'll add the comment to Kube.ts for now and we can decide when we need this.